broadinstitute/single_cell_portal_core
# Instant header validation for local AnnData files (SCP-5718)

This helps users avoid waiting 15-45 minutes to upload a big H5AD file, only to learn of a user error at the end.

### Impact
Previously, authors would learn of any validation errors in their AnnData file only after uploading it and then checking their email.  That's quite slow because .h5ad files in single cell science tend to be large: 2 GB, 6 GB, sometimes 60 GB.  Needing to upload such big files to detect validation problems is frustrating.

Now, some problems are found _before_ upload.  Basic AnnData validation is done right in the web upload UI in less than a second.  It validates AnnData metadata headers with a few existing rules, and reports any problems instantly and in context.  The user journey is smoother, and thus faster.  

We made AnnData upload generally available in July 2023, and have noticed a distinct uptick in its usage in the last few months.  So improving this UX seems worthwhile.

### Demo
Here's how it looks!

https://github.com/user-attachments/assets/7f21c8d4-33f5-4edf-94f6-c43193feb3bb

### Technical notes
This uses [hdf5-indexed-reader](https://github.com/jrobinso/hdf5-indexed-reader)*, an efficient JavaScript library for HDF5 files.  JS in the web browser parses the start of the AnnData file, reads `obs` keys, and applies some basic pre-existing [client-side file validation](https://github.com/broadinstitute/single_cell_portal_core/pull/1116) (CSFV) rules.  It reuses existing rules via minor rearrangement of parsed AnnData.  So we can use the same validation rule code for both classic and AnnData files.

Parsing is streaming and non-blocking.  That's important because AnnData files can be larger than client memory.  We recognized this as an underlying blocker for AnnData back in November 2022, and [started a community discussion about how to solve it](https://github.com/usnistgov/h5wasm/issues/40).  Since then, building on @bmaranville's foundational work, @jrobinso made `hdf5-indexed-reader` and we now have a library that neatly fits this need.

This changeset also includes a standalone HTML page and spinning DNA graphic.  That enables nimbly experimenting with HDF5 and JS using something like [Simple Web Server](https://simplewebserver.org/).

\* Because [hdf5-indexed-reader](https://www.npmjs.com/package/hdf5-indexed-reader) isn't on NPM, it can't be used as a conventional package.  So I went ahead and published an org-scoped version of it, [@single-cell-portal/hdf5-indexed-reader](https://www.npmjs.com/package/@single-cell-portal/hdf5-indexed-reader).  The [package.json source](https://github.com/broadinstitute/hdf5-indexed-reader/tree/ew-npm-scoped-package) trivially adds a scope to the original name.  This ensures the canonical version can use a simpler top-level name (presumably `hdf5-indexed-reader`) if / when it's published to NPM.

### Next steps
* Instant header validation for _remote_ AnnData files
* Instant _content_ validation for AnnData files

### Test
A new automated test confirms some behavior.  To manually test:
1.  `yarn install`
2. Ensure Vite frontend service worker cache is disabled
3.  Create a new study "AnnData CSFV"
4.  Select AnnData upload UX
5.  In AnnData tab, click "Choose file"
6.  In Finder, go to `single_cell_portal_core/test/test_data/`
7.  Select `anndata_test_bad_header_no_species.h5ad`
8.  Confirm error message is shown, including "File is missing required obs keys: species"
9.  Click "Delete" to remove errant file, then click "Choose file"
10.  In same directory as before, select `anndata_test.h5ad`
11.  Confirm no error message is shown, and "Save" button is enabled
12. Upload `anndata_test_bad_header_no_species.h5ad` and `anndata_test.h5ad` to your study's GCP bucket
13. Click "Use bucket path"
14. Enter "anndata_test_bad_header_no_species.h5ad" in text input
15. Click off input to blur / change browser focus
16. Confirm no error message is shown (remote AnnData CSFV will be added next) and "Save" button is enabled

### Further details
For more context, see [2024-08-13 SCP demo video](https://drive.google.com/file/d/12t_AgcHRknZepvsRw2fDrF8PS9vWq0cP/view?t=40m28s).  This satisfies SCP-5718.
